File: BM30_二叉搜索树与双向链表/python_solutions/recursion_solution.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

# 
# @param pRootOfTree TreeNode类 
# @return TreeNode类
#
class Solution:

    def __init__(self):
        self.head = None
        self.pre = None
    
    def Convert(self, pRootOfTree):
        if not pRootOfTree: return pRootOfTree

        self.Convert(pRootOfTree.left)
        
        if not self.head:
            self.head = pRootOfTree
            self.pre = pRootOfTree
        else:
            self.pre.right = pRootOfTree
            pRootOfTree.left = self.pre
            self.pre = pRootOfTree

        self.Convert(pRootOfTree.right)
        
        return self.head

File: BM30_二叉搜索树与双向链表/javascript_solutions/RecursionSolution.ts
class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.left = (left === undefined ? null : left)
        this.right = (right === undefined ? null : right)
    }
}

let head: TreeNode | null | undefined = null;
let pre: TreeNode | null | undefined = null;

export function Convert(pRootOfTree: TreeNode | null): TreeNode | null | undefined {
    if (!pRootOfTree) return null;

    Convert(pRootOfTree.left);

    if (!head && !pre) {
        head = pRootOfTree;
        pre = pRootOfTree;
    } else {
        if (!pre) throw new Error("神学事件");
        pre.right = pRootOfTree;
        pRootOfTree.left = pre;

        pre = pRootOfTree;
    }

    Convert(pRootOfTree.right);

    return head;
}